### DISCUSSION TITLE
[AsyncNotifier] How to set AsyncError in custom state?

### DISCUSSION THREAD

**Discussion by devnta** [rrousselGit/riverpod#3361] [Q&A] [answered]
Assume I have `MovieState` using freezed:
```
@freezed
class MovieState with _$MovieState {
  const factory MovieState({
    @Default([]) List<MovieResponse> movies,
  }) = _MovieState;
}

```
In `MovieNotifier`
```
class MovieNotifier extends AsyncNotifier<MovieState> {
@override
  Future<MovieState> build() async {
    state = const AsyncLoading();
    final repository = ref.read(movieRepositoryProvider);
    final response = await repository.fetchMovies();
    List<MovieResponse> result = [];
    response.fold((l) {
      state = AsyncError(l, StackTrace.current);
    }, (r) {
      result = r;
      state = const AsyncData(MovieState());
    });
    return MovieState(
      movies: result,
    );
  }
}
```
In `MovieView`:
```
asyncValue.when(
        data: (data) => ListView.builder(
          itemBuilder: (context, position) {
            return MovieItem(movie: data.movies[position]);
          },
          itemCount: data.movies.length,
        ),
        error: (error, trace) => const SizedBox(),
          child: CircularProgressIndicator(),
        ),
      )
```
When `hasError` is true and `state = AsyncError('My error', StackTrace.current); ` triggered, `error` on `asyncValue.when` not called, the `data` always called.

## Question:
**1. How to trigger `state = AsyncError in AsyncNotifier?`**
**2. What I should return for each case success and error? How to use `copyWith` function like `StateNotifier`**

**Comment by noriHanda** [comment]
Your build method should be changed like this
```dart
Future<MovieState> build() async {
    final repository = ref.watch(movieRepositoryProvider);
    final response = await repository.fetchMovies();
    return MovieState(movies: response);
  }
```
Loading and error associated with Future calls gets automatically handled inside build method of AsyncNotifier

**Reply by devnta** [reply]
That means I no need use `state = AsyncLoading` and `state = AsyncError` to handling error and loading?
What should i return when there is error on the call `final response = await repository.fetchMovies();`?
Because if fetchMovies error then response is null.
And how to I set state = AsyncError so that MovieView widget display error widget?

**Reply by snapsl** [reply]
Yes, you do not need to set `AsyncLoading` and `AsyncError` manually. During `build` time of the provider, where the data is loaded, `AsyncValue` is set to `AsyncLoading` automatically. 
If `repository.fetchMovies()` fails, this will be handled by riverpod using the `error` case. You can test this by manually throwing an `Exception` in `repository.fetchMovies()`.

**Reply by devnta** [reply]
@snapsl I was manually throwing an Exception in repository.fetchMovies(). Because `build` method always return `MovieState`, `when` method always triggered `data`.

**Reply by devnta** [reply]
@snapsl 
If I throwing on `build` method:
```
Future<MovieState> build() async {
    final repository = ref.watch(movieRepositoryProvider);
    final response = await repository.fetchMovies();
    if(success){
    return response;
    }else{
    throw Exception();
    }
  }
 
  ```
  the `error` will triggered.
  Is this the best practice?

**Reply by snapsl** [reply]
This should not happen if you implement your build as suggested by @noriHanda.

**Reply by devnta** [reply]
@snapsl If I implement as suggested by @noriHanda , when.data always triggered.
My problem is how to triggered when.error?

**Reply by noriHanda** [reply]
@phamquoctrongnta repository.fetchMovies() might be the cause. Can you share the code?

**Reply by devnta** [reply]
@noriHanda 
My repository.fetchMovies():
```
Future<Either<Exception, List<MovieResponse>>> fetchMovies() async {
    final dio = Dio();
    const url = Endpoint.getTopRatedMovies;
    try {
      dio.interceptors.add(AddHeaderInterceptor());
      final response = await dio.get(url);
      final mapFromJson = response.data['results'] as List;
      final result = mapFromJson.map((e) {
        return MovieResponse.fromJson(e);
      }).toList();
      return Right(result);
    } on Exception catch (e) {
      return Left(e);
    }
  }
```

**Reply by noriHanda** [reply]
@phamquoctrongnta yeah... you are returning a value no matter the error, so that's the cause. You need to `throw` instead of `return` when error.

**Reply by devnta** [reply]
@noriHanda 
Thanks for response.
```
Future<MovieState> build() async {
    final repository = ref.watch(movieRepositoryProvider);
    final response = await repository.fetchMovies();
    return MovieState(movies: response);
  }
```
Can I throwing the exception in `build` method instead `repository.fetchMovies()`?
Is the best practice?

**Reply by noriHanda** [reply]
@phamquoctrongnta It's not the best practice because the exception comes from the fetchMovies function but it should work.

**Reply by noriHanda** [reply]
@phamquoctrongnta Please mark this discussion answered once you think is done :)

**Reply by devnta** [reply]
@noriHanda Sorry for the late.
If I want to use `Either` class, I can't throw an exception on `repository.fetchMovies()` . In that case, where do I throw?

**Reply by noriHanda** [reply]
@phamquoctrongnta I don't see the need of `Either` because you can just `throw Exception(e)`. That `e` becomes `e` of `error: (e, s) => `of `when` method.

**Reply by utamori** [reply]
You should throw when it's a Left type.
[examples/pokeapi_functional/lib/controllers/pokemon_provider.dart](https://github.com/SandroMaglione/fpdart/blob/d1f5f1531ac069d6d67280991fc45801deb463da/examples/pokeapi_functional/lib/controllers/pokemon_provider.dart)

**Reply by devnta** [reply]
@noriHanda I used `Either` because I want to catch the error clearly.
Ex:
```
if(left){
doError();
}else{
doSomething();
}
```

**Reply by devnta** [reply]
@utamori Thanks for answer.
That means is I should throw an error on `build` method of `Notifier`, right?

**Reply by devnta** [reply]
@utamori 
In 28 line: `(error) => AsyncError(error, StackTrace.current),`
I think it won't fire on `when.error` because the function always returns `unit`. Have you tested it?

**Reply by noriHanda** [reply]
@phamquoctrongnta I still don't get it... You can just see if the state `hasError`.

**Reply by devnta** [reply]
@noriHanda 
You said:

> I don't see the need of Either because you can just throw Exception(e).

I reply:

> I used Either because I want to catch the error clearly.
Ex:
if(left){
doError();
}else{
doSomething();
}


I want to return `Either` with `Left` instead throw an `Exception` on repository.fetchMovies().
So can I throw an exception on `build` method of Notifier? Is it the best practice?
Sorry for my English, hope you understand.

**Reply by noriHanda** [reply, marked as answer]
@phamquoctrongnta Your English is good. English is not my first language either so we are on the same side :)
What I was trying to tell you was that `throw`ing `Exception` was a valid way of returning details of error. `throw` is more suitable than `return` for exceptions because IDE can jump to where exception has occurred, which means it's more developer friendly. If you just return `Exception` as you mention, you need to handle manually inside `build` method of Notifier and it gets harder to track where the `Exception` originates from.

**Reply by devnta** [reply]
@noriHanda Thanks for detail answer.
I'm clean.
Thank you, again.

**Reply by devnta** [reply]
@noriHanda Sorry.
Do you know use `copyWith` method in `AsyncNotifier`?
Ex:
```
state = state.copyWith(
a: newA,
)
```

**Reply by noriHanda** [reply]
@phamquoctrongnta There's no `copyWith` method in `AsyncValue`. There is, however, `copyWithPrevious` that is used internally. There isn't many valid use cases for `copyWithPrevious`, so it shouldn't be your first option.
What you should think about using first is `AsyncValue.guard`. Any asynchronous operations inside guard clause is handled properly.
https://pub.dev/documentation/riverpod/latest/riverpod/AsyncValue/guard.html

**Reply by devnta** [reply]
@noriHanda 
If I have complex state, what should I do to update new state based on old state?
Ex: I have `MyState` have 2 flag: `bool isChecked = false`, `bool isLogin = false`.
Assume click a any button to change `isLogin` to `true`.
Then click a another button to change `isChecked` to `true` but keep `isLogin` state is `true`?

**Comment by Trung15010802** [comment]
> @noriHanda If I have complex state, what should I do to update new state based on old state? Ex: I have `MyState` have 2 flag: `bool isChecked = false`, `bool isLogin = false`. Assume click a any button to change `isLogin` to `true`. Then click a another button to change `isChecked` to `true` but keep `isLogin` state is `true`?

You can do like me. use state.value
![image](https://github.com/rrousselGit/riverpod/assets/90593296/f5f248c0-e0d9-47fc-b40b-1e4a1fd3724e)

**Reply by devnta** [reply]
Thanks for response, bro.
